### PR TITLE
drivers: flash : Cadence nand driver

### PR DESCRIPTION
**Pull request by nbalabak**
Implemented the Cadence NAND driver with the following features:

- Added new driver for Cadence Nand
- Supports 3 modes : CDMA , PIO, generic mode
- Support Operation : read, write, erase
- Supports x8 and x16
- Supports SDR datarate

### CODE REVIEW

**Comment by nbalabak** [comment]
Hello @de-nordic ,
               could you please run compliance check and scan code for license?

**Comment by nbalabak** [comment]
Hello @de-nordic ,
could you please run compliance check and scan code for license?

**Comment by nbalabak** [comment]
@kartben @SgrrZhf @povergoing could you please review nand driver?

**Review by povergoing — COMMENTED** [review]
LGTM overall, but I am not a nand expert, only some trivial comments

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:1280]: This blocks keep on repeating.
Wouldn't it look better if function would be defined, maybe inline, that would change depending on the config defined and called here?

File: drivers/flash/Kconfig.cadence_nand
Comment (nbalabak) [reply drivers/flash/Kconfig.cadence_nand]: @povergoing i updated changed

File: drivers/flash/Kconfig.cadence_nand
Comment (nbalabak) [reply drivers/flash/Kconfig.cadence_nand]: @povergoing i updated changed

File: drivers/flash/Kconfig.cadence_nand
Comment (nbalabak) [reply drivers/flash/Kconfig.cadence_nand]: @povergoing i updated changes

File: CODEOWNERS
Comment (nbalabak) [reply CODEOWNERS]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i created inline function and updated code

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:80]: It does not seem you are querying this information from external device, so I guess you could just pre-calculate this at init, instead of doing that for each layout request.

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:128]: Same as in erase

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:154]: same as in erase.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:497]: Why just not define single cdsn_transfer_complete, with `struct cadence_nand_params *params` and do `ARG_UNUSED(params)` when not used?

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: @de-nordic as per review comments changes updated

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: @de-nordic as per review comments changes updated

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: @de-nordic as per review comments changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic i updated changes

File: drivers/flash/flash_cadence_nand_ll.h
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.h:368]: @de-nordic these api exposed to upper layer. we made two layers flash_cadence_nand_ll.c and  flash_cadence_nand.c.
 these function exposed to  flash_cadence_nand.c .

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c:284]: This code is not repeated; its behavior changes based on the SDR and DDR mode register values. The data rate is extracted from the dtsi, but there may be multiple nodes, some with SDR and others with DDR. Therefore, conditional compilation based on these scenarios is not feasible, in my opinion. Do you have any suggestions? Please share your insights

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c:194]: if only F_CSTAT_COMP bit then we can continue WAIT_FOR loop. but once complete flag set the possiblity there read/write operation fail bit also to set (line 195 and 196). did you have any sugesstion please let me know

File: drivers/flash/Kconfig.cadence_nand
Comment (nbalabak) [reply drivers/flash/Kconfig.cadence_nand]: @povergoing  Can you please review the changes I made based on your comments?

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: > Looks like significant portion of code here repeats.

This code is not repeated; its behavior changes based on the SDR and DDR mode register values. The data rate is extracted from the dtsi, but there may be multiple nodes, some with SDR and others with DDR. Therefore, conditional compilation based on these scenarios is not feasible, in my opinion. Do you have any suggestions? Please share your insights

Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: > `opr_mode` check that could result in `-ENOTSUP`, and `opr_mode` check does not depend on run-time info from dev, should happen somewhere above the line 222 as there are already a lot of ops called on a device before deciding that the driver will not support the op anyway. I can also see that this is internal utility function and `opr_mode`, when this function gets called, comes from `cadence_nand_params->datarate_mode`, which is initialized at compile time from DTS. This means that the valid data rate mode could probably be checked by driver init, against allowed values, or verified at compile time. Is it possible to pass incorrect value here?

@de-nordic changes updated. datarate_mode value verified in dtsi via yml file not possible to get -ENOTSUP so removed

File: dts/bindings/flash_controller/cdns,nand.yaml
Comment (nbalabak) [reply dts/bindings/flash_controller/cdns,nand.yaml:13]: updated with enum values

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: removed else if and changes to else

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: changes updated

**Comment by nbalabak** [comment]
@de-nordic could you please review PR i fixed your review comments?

**Comment by nbalabak** [comment]
> @de-nordic could you please review PR i fixed your review comments?

Hello @de-nordic,
       could you please review PR i fixed your review comments?

**Comment by nbalabak** [comment]
> @povergoing could you please review PR i fixed your review comment?

@povergoing could you please review PR i fixed your review comment?

**Comment by nbalabak** [comment]
@kartben could you please review this PR?

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:93]: Just do nothing, I think that is what most other drivers do.
I do not think that asking to erase 0 size is a failure, nothing really happens.

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:102]: I think it is NAND not Nand, and you could log the error here, in case it is -EINVAL that is coming somewhere from `cdns_nand_erase`, to distinguish from the one in line 92 . Bang is not needed, people are probably already scared by the "E:" at the beginning of error message anyway.

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:91]: I think this message is pointless. This is single point in `flash_cdns_nand_erase` where you can get the -EINVAL.
I think the `LOG_ERR` should be reserved for something that is coming from a lower level, that can not be easily identified.
I am quite sure that this can be easily figured out by user without the message.

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand.c:206]: It is NAND.

**Review by de-nordic — COMMENTED** [review]
Still reviewing...

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:652]: Pointless, just return `ret` at the end.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:644]: As far as I can tell, this could be done inside `cdns_cdma_desc_transfer_finish`.
Correct me if I am wrong but the first param to `cdma_allocate_desc_memory` is always third param to `cdns_cdma_desc_transfer_finish` and second param to the former is always fourth param to the later.
Wouldn't it save flash to move the block above to the `cdns_cdma_desc_transfer_finish` ?

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:688]: same as in `cdns_nand_cdma_write`

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:722]: Same as in `cdns_nand_cdma_write`

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:771]: You could just have some `INT_SEM_TAKE` macro defined, that would depend on the `CONFIG_CDNS_NAND_INTERRUPT_SUPPORT`... less ifdefs cutting into code.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:570]: IMHO zeroing buffer at allocation is a waste of time, unless hardware assisted, because most of the buffer will be filled in with some stuff. Unless these buffers are passed to something that always requires it zeroed, then fine.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:686]: There is double free everywhere where you do that after calling `cdns_cdma_desc_transfer_finish`, because on error conditions in `cdns_cdma_desc_transfer_finish` you release the buffer first time.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:1146]: I guess when `ret` is error you keep on looping?

File: drivers/flash/flash_cadence_nand_ll.c
Comment (de-nordic) [inline drivers/flash/flash_cadence_nand_ll.c:1023]: Not insisting but you could probably have just copyin/copyout fuctions for that.
I doubt we save any flash by stacking both directions here, assuming that the dir has to be passed to a function which means that each call has to pass it through some register, so probably save contents of the register on stack, and so on.

File: drivers/flash/flash_cadence_nand.c
Comment (de-nordic) [reply drivers/flash/flash_cadence_nand.c:115]: Now I see that the len is even checked in ll functions. Isn't there too much checking done, I mean on too many levels?

Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: removed from .c file len checking condition. Now ll function if len = 0 means it will return 0.

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: removed this condition.

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c:96]: yes its in the ret range

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: removed this log error.

File: drivers/flash/flash_cadence_nand.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand.c]: chnages updated.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: Changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: removed this condition

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: moved cdma_allocate_desc_memory functionality to cdns_cdma_desc_transfer_finish block.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: Changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: Changes updated

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: macro added and replaced with new macro.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: removed.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c:1086]: added return condition if error occurs.

File: drivers/flash/flash_cadence_nand_ll.c
Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: copyin and copyout two seprate function added.

Comment (nbalabak) [reply drivers/flash/flash_cadence_nand_ll.c]: @de-nordic  i fixed all review comments could you please review once?

**Comment by nbalabak** [comment]
@kartben @SgrrZhf @povergoing could you please review nand driver?

**Review by de-nordic — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
LGTM

**Comment by nbalabak** [comment]
@de-nordic  could you merge my PR?

**Comment by kartben** [comment]
> @de-nordic could you merge the PR?

One of our release engineers will probably pick it up shortly now that it has all required approvals.

**Comment by nbalabak** [comment]
> > @de-nordic could you merge the PR?
> 
> One of our release engineers will probably pick it up shortly now that it has all required approvals.

@kartben Thank you for sharing information.

**Comment by de-nordic** [comment]
> @de-nordic could you merge the PR?

I need to take my merge button for repair because it is gray and broken all the time. You will need to wait for release engineers to step in.

**Pull request merged**
